internal/olm/operator/bundle/install.go

```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+			isChannelPresent = true
+			break
+		}
+	}
+
+	// insert bundle to index if package/bundle/channel is not present on the index;
+	// and insert only bundle declarative config channel to index when both bundle/package are present but not the channel.
+	if !isPackagePresent && !isBundlePresent && !isChannelPresent {
+		imageDeclConfig.Bundles = append(imageDeclConfig.Bundles, bundleDeclConfig.Bundles[0])
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+		imageDeclConfig.Packages = append(imageDeclConfig.Packages, bundleDeclConfig.Packages[0])
+		imageDeclConfig.Others = append(imageDeclConfig.Others, bundleDeclConfig.Others...)
+	} else if isPackagePresent && isBundlePresent && !isChannelPresent {
+		// if the bundle already exists in a different channel,
+		// we already have the bundle and package blobs in the FBC, in which case,
+		// we need to reference it in the channel we want it to be in.
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+	} else if isBundlePresent {
+		return nil, fmt.Errorf("bundle %q already present in the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	} else {
+		return nil, fmt.Errorf("cannot add bundle %q to the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	}
+
+	log.Infof("Inserted the new bundle %q into the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+
+	return imageDeclConfig, nil
+}
+
+// createFBC generates an FBC by creating bundle, package and channel blobs.
+func (f *FBCContext) createFBC(ctx context.Context) (bundleDeclcfg, error) {
+	var bundleDC bundleDeclcfg
+	// Rendering the bundle image into a declarative config format.
+	log.SetOutput(ioutil.Discard)
+	render := action.Render{
+		Refs: f.Refs,
+	}
+
+	// generate bundles by rendering the bundle objects.
+	declcfg, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+	if err != nil {
+		return bundleDeclcfg{}, fmt.Errorf("error rendering the bundle image: %v", err)
+	}
+
+	// Ensuring a valid bundle size.
+	if len(declcfg.Bundles) != 1 {
+		return bundleDeclcfg{}, fmt.Errorf("bundle image should contain exactly one bundle blob")
+	}
+
+	bundleDC.Bundle = declcfg.Bundles[0]
+
+	// generate package.
+	bundleDC.Package = declarativeconfig.Package{
+		Schema:         schemaPackage,
+		Name:           f.Package,
+		DefaultChannel: f.ChannelName,
+	}
+
+	// generate channel.
+	bundleDC.Channel = declarativeconfig.Channel{
+		Schema:  schemaChannel,
+		Name:    f.ChannelName,
+		Package: f.Package,
+		Entries: []declarativeconfig.ChannelEntry{f.ChannelEntry},
+	}
+
+	return bundleDC, nil
+}
+
+// stringifyDecConfig writes the generated declarative config to a string.
+func stringifyDecConfig(declcfg *declarativeconfig.DeclarativeConfig) (string, error) {
+	var buf bytes.Buffer
+	err := declarativeconfig.WriteJSON(*declcfg, &buf)
```

Comment:
Maybe consider `YAML` instead? I think YAML is generally smaller than JSON. So we'd be less likely to run into the etcd size limit issues I mentioned before.

Reply:
Sounds good, I think that will partially fix the size limit issues. I'll change it to encode to YAML instead.